components/script/dom/htmlimageelement.rs

```diff
@@ -871,6 +917,175 @@ impl HTMLImageElement {
         ScriptThread::await_stable_state(Microtask::ImageElement(task));
     }
 
+    /// <https://html.spec.whatwg.org/multipage/#img-environment-changes>
+    pub fn react_to_environment_changes(&self) {
+        // Step 1
+        let task = ImageElementMicrotask::EnvironmentChangesTask {
+            elem: DomRoot::from_ref(self),
+            generation: self.generation.get(),
+        };
+        ScriptThread::await_stable_state(Microtask::ImageElement(task));
+    }
+
+    /// Step 2-12 of https://html.spec.whatwg.org/multipage/#img-environment-changes
+    fn react_to_environment_changes_sync_steps(&self, generation: u32) {
+        fn add_cache_listener_for_element(image_cache: Arc<ImageCache>,
```

Comment:
Let's add a TODO comment about reducing the duplicating between this code and the code for regular image fetching. We can file a followup issue to address that.